### ISSUE TITLE
Tsconfig option to disallow features requiring transformations which are not supported by Node.js' --strip-types

### ISSUE THREAD

**Issue by marcomuser** [microsoft/TypeScript#59601]
### 🔍 Search Terms

--strip-types

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

Node.js has introduced an experimental flag that allows type annotations to be stripped. However, since Node.js only erases inline types, all TypeScript features that involve replacing TypeScript syntax with new JavaScript syntax will fail as described in the [Node.js docs](https://nodejs.org/api/typescript.html#unsupported-typescript-features). The following features are listed in the docs as the most important features not supported:
- Enum
- experimentalDecorators
- namespaces
- parameter properties

Would it be possible to introduce a single flag in tsconfig that tells the compiler in one fell swoop that all these features should not be enabled to ensure compatibility with Node.js `--strip-types`?

### 📃 Motivating Example

If there was such a configuration option, you could easily ensure that the code you write always contains only standard JavaScript + type annotations that can be executed by Node.js without installing any additional packages.

### 💻 Use Cases

Finding the correct configuration of tsconfig for different Node.js projects is already relatively complicated. A simplified configuration that allows you to author compliant Typescript code that works smoothly with Node.js new out-of-the-box ts support via the `--strip-types` flag would be a great help.

**Comment by RyanCavanaugh** [comment]
[#54283 ](https://github.com/nodejs/node/pull/54283) is relevant here

**Comment by Bnaya** [comment]
Feels like a lint rule and not a ts flag

I have a strong feeling that node will eventually support everything under `isolatedModules`, as it's the baseline for transformers these day and there you got your flag :)

**Comment by RyanCavanaugh** [comment]
Discussed a bit with some folks internally and there was possible appetite for this since. This has been a longstanding request for other reasons (ideologically purity [complimentary], future-proofing, de facto tool support, etc). A sticking point is what the heck to name it and some suggestions to get the ball rolling would be useful.

**Comment by marcomuser** [comment]
Wohooo! 🥳 So if you feel like poking a little fun at the ideological purity section, call it `ecmastrict`. 😁

**Comment by robpalme** [comment]
Just to add to @RyanCavanaugh's list of reasons to add this mode:  a further benefit not yet stated is that it will permit TypeScript (if the team wishes) to introduce a new JS emit mode that preserves JS syntax coordinates, meaning no sourcemap is required.

SWC have [already shipped](https://swc.rs/docs/references/wasm-typescript) such an emitter written in Rust and compiled to Wasm.  @acutmore will soon be open-sourcing another example of such an emitter - this time written in TypeScript.  [`ts-blank-space`](https://gist.github.com/acutmore/27444a9dbfa515a10b25f0d4707b5ea2) is a type-stripper built on top of the TypeScript parser.  It is ~700 lines of code.  With large files it achieves a speed up of 4.7x relative to TS 5.5 `ts.transpileModule` with `noCheck`. With small files it goes even faster due to less GC.

I agree the hardest problem is what to name it.

**Comment by RyanCavanaugh** [comment]
I'll just start throwing ideas in:
 * `--noTranspiledFeatures`
 * `--typeSyntaxOnly`
 * `--disallowRuntimeSyntax`

Note that we almost always prefer a flag like this to be `false` by default, so the name should reflect that

**Comment by marcomuser** [comment]
I like the direction that `--typeSyntaxOnly` is taking, but given the waves that the level 1 proposal has made, I would change it slightly to match its title: `--typeAnnotationsOnly`.

**Comment by allisonkarlitskaya** [comment]
Highly relevant:
 - https://github.com/tc39/proposal-type-annotations#proposal
 - https://tc39.es/proposal-type-annotations/

> No, not all of today's TypeScript syntax would be supported by this proposal. This is similar to how Babel support for TypeScript does not support all of the existing TypeScript syntax.
>
> For example enums, namespaces and class parameter properties are unlikely to be supported. In addition, specifying type arguments at function call-sites will require a slightly different syntax.

This is early stuff, but it seems like this would probably be the thing to target with such an option.

**Comment by RyanCavanaugh** [comment]
`--typeSyntaxOnly` 👍

Rough notes

Who is this for?
 * Ideological purists
 * Space-only transpilation users
 * Today's version of node (but not tomorrow's?)
 * Forward-versioning safety if committee changes its mind about enum
 * People who don't like adding a linter
 * Keep syntax in the space that's *likely* to be support by type annotations in JS, if that ever happens

Downsides: nothing allowed in .ts exactly replicates what `enum` does today

Must use `verbatimModuleSyntax` and `isolatedModules` to turn this on

Applies only to .ts, not .d.ts

Recommended to combine with `verbatimModuleSyntax` and `isolatedModules`, but not required

Exact definition of what's disallowed and what's not
```ts
import x = require('fs'); // no (CJS+VMS will not have a good workaround at this time)
import A = e.p; // no
class X {
  public x; // OK (just erase `public`)
  constructor(public y) { } // not OK - runtime-observable
}
enum X { } // All forms (including `const`) not OK
namespace T { } // OK (type-only)
namespace X { // Not OK (instantiated)
  const x = 1;
}
```

**Comment by robpalme** [comment]
Thanks for the clear concise comprehensive update, @RyanCavanaugh.

```ts
namespace T { } // OK (type-only)
```

This was the only surprise to me.  I appreciate it does not emit anything so can be considered erasable.  I'm curious why anyone would use this form and whether its used in real-life.

**Comment by RyanCavanaugh** [comment]
Non-instantiated namespaces are need to do certain kinds of declaration merging, e.g. adding a no-emit `static` member to a class

**Comment by Conaclos** [comment]
> `namespace T { } // OK (type-only)`

I still find confusing that `declare` is not required here to make the namespace ambient/non-instantiated.
This requires extra work for a transpiler/linter to check whether the namespace is ambient.
I am not sure if node experimental strip type performs this check or just reject `namespqace` without `declare`.

> `const foo = {...} as enum`
>
> Wouldn't be a full substitute.
>
> ```ts
> const x = {
>   a: 1,
>   b: a, // can't do this
> } as enum;
> ```

I really like this proposal from this [design notes](https://github.com/microsoft/TypeScript/issues/59658).
Yes it is not a substitute, however it allows most of the cases users encounter and this provides a concise syntax for users that want to avoid runtime TS features.

**Comment by jakebailey** [comment]
`declare namespace` is not a replacement for type-only namespaces; you can declare any value-space thing and it will "just work". Code like:

```ts
declare namespace T {
    export function doSomething(): void;
}

console.log(T.doSomething())
```

Will typecheck, but then crash at runtime.

This syntax is designed for declaring things that already exist in the runtime environment (think `declare var __WEBPACK_REQUIRE__: any`), which is a generally useful thing, but isn't a safe replacement.

Compare that to a rule which bans "instantiated namespaces", complaining about the value declaration.

**Comment by robpalme** [comment]
The [ts-blank-space](https://bloomberg.github.io/ts-blank-space/) compiler I mentioned [earlier in this issue](https://github.com/microsoft/TypeScript/issues/59601#issuecomment-2286741349) is now publicly available.

One correction to the earlier post is that the performance multiplier relative to classic `ts.transpileModule` with `noCheck` was previously 4.7x but after further optimization is now 5.6x.  The full performance results including benchmarks are [here.](https://github.com/bloomberg/ts-blank-space/tree/main/perf)

**Comment by khaosdoctor** [comment]
> The next step appears to be naming the flag. So here are some options.
> 
> (Personally I mildly prefer `noTypeDrivenEmit`)
> 
> # Node-centric
> * `stripTypes: true`
> * `stripTypesOnly: true`
> * `onlyStripTypes: true`
> * `justStripTypes: true`
> * `enforceStripTypes: true`
> * `(enforce/match)Node: true`
> * `(enforce/match)NodeRestrictions: true`
> 
> # Opt-in Positives
> * `erasableTypes: true`
> * `erasableTypesOnly: true`
> * `onlyErasableTypes: true`
> * `eraseTypes: true`
> * `typesOnly: true`
> * `onlyTypes: true`
> * `typeAnnotationsOnly: true`
> * `onlyTypeAnnotations: true`
> * `jsPlusTypes: true`
> * `verbatimCode: true`
> * `verbatimEmit: true`
> * `verbatimJavaScript: true`
> 
> # Opt-out Negatives
> All these could either be inverted with a `no` or `prevent` prefix, or could simply default to `true` without a prefix.
> 
> * `(TypeScript)RuntimeFeatures: false`
> * `(TypeScript)CodeGen: false`
> * `TypeDrivenEmit: false`
> * `EnumsNamespacesOrParameterProperties: false`
> * `Regrats: false`

I'm personally more into an option that's more descriptive. So from these options I'd go for:

While these two makes me think of "typescript will only strip the types" and not "I can't use namespaces" I think it's a good name:
* `stripTypesOnly: true`
* `onlyStripTypes: true`

Those I think are more descriptive to what it's actually going to do:
* `erasableTypesOnly: true`
* `onlyTypeAnnotations: true`
* `(TypeScript)CodeGen: false`
* `EnumsNamespacesOrParameterProperties: false`

**Comment by jakebailey** [comment]
Anders had previously suggested `--erasableSyntaxOnly` (which I like), similar to `--erasableTypesOnly` ish.

I do not think that naming this such that it's related to what Node is doing is a good idea, since it's entirely possible Node decides to instead enable _all_ features supported with isolatedModules (e.g. enums, namespaces), so if we had to change our behavior to match, that would give people who _actually_ want to disable these features for different reasons from having a flag at all.

**Comment by allisonkarlitskaya** [comment]
> I do not think that naming this such that it's related to what Node is doing is a good idea, since it's entirely possible Node decides to instead enable all features...

Indeed, in addition to `--experimental-strip-types` they now also have `--experimental-transform-types`.

**Comment by khaosdoctor** [comment]
> > I do not think that naming this such that it's related to what Node is doing is a good idea, since it's entirely possible Node decides to instead enable all features...
> 
> Indeed, in addition to `--experimental-strip-types` they now also have `--experimental-transform-types`.

And one implies the other too, so if you use `--experimental-transform-types` the `--experimental-strip-types` will be implied because it's also needed. I agree with @jakebailey on this, maybe using something akin to `erasableTypesOnly` is better

**Comment by RyanCavanaugh** [comment]
Absent a name everyone likes, we found a name no one objected to: `erasableSyntaxOnly`

If anyone wants to send a PR for this in the next week or two LMK, otherwise we'll implement.

**Comment by justinfagnani** [comment]
Since `experimentalDecorators` has it's own flag, it probably doesn't to be covered in `erasableSyntaxOnly`. If someone writes a config with both `erasableSyntaxOnly: true` and `experimentalDecorators: true` then they probably meant that.

**Comment by ljharb** [comment]
Given that configs can be extending other ones, I'm not sure one can assume it was intended.

**Comment by justinfagnani** [comment]
@danfo those options names don't make sense to me because only decorators was ever experimental - the others are enabled by default and not opt-in.

**Comment by RyanCavanaugh** [comment]
#61011 is merged, which will be in 5.8 beta. Happy coding!

**Comment by aceArt-GmbH** [comment]
Wouldn't it be better to add some api for Node.js, Bun, esbuild and all the others to allow for actual TypeScript support with typechecking instead of adding more "Typescript support but not really"? 🙃

**Comment by sosoba** [comment]
What about class decorators and explicit resource management? Both requires transpiling but `--erasableSyntaxOnly` does not control their use. [Example](https://www.typescriptlang.org/play/?noEmitHelpers=true&target=11&module=200&erasableSyntaxOnly=true&ssl=10&ssc=1&pln=1&pc=1&importHelpers=true#code/KYDwDg9gTgLgBAYwgOwM7wCbCVAhjaOAXjgAoEAbVALjgFdkBrZCAd2QBpEUZQZaAwhVypUAEWzR80ATz4BKYgD44AbwC+AbgCwAKD0ABLDmlQ9lEajgC1e9Xr2q9cF-VQBLZAHNExOMmBWa1J5HV17XSA):

```ts
export const decorator = (cls: unknown, context: ClassDecoratorContext) => {};

@decorator
class C {
}

{
    using c = new C();
}
```

**Comment by marco-ippolito** [comment]
> What about class decorators and explicit resource management? Both requires transpiling but `--erasableSyntaxOnly` does not control their use. [Example](https://www.typescriptlang.org/play/?noEmitHelpers=true&target=11&module=200&erasableSyntaxOnly=true&ssl=10&ssc=1&pln=1&pc=1&importHelpers=true#code/KYDwDg9gTgLgBAYwgOwM7wCbCVAhjaOAXjgAoEAbVALjgFdkBrZCAd2QBpEUZQZaAwhVypUAEWzR80ATz4BKYgD44AbwC+AbgCwAKD0ABLDmlQ9lEajgC1e9Xr2q9cF-VQBLZAHNExOMmBWa1J5HV17XSA):
> 
> ```ts
> export const decorator = (cls: unknown, context: ClassDecoratorContext) => {};
> 
> @decorator
> class C {
> }
> 
> {
>     using c = new C();
> }
> ```
> 

The syntax that is coming into the language (decoratora, using etc...) is not banned, but simply Node.js javascript engine does not supportbit yet
